### PR TITLE
adding visual preview window

### PR DESCRIPTION
**Pull request by hfazai**

### CODE REVIEW

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 1]
```diff
@@ -143,7 +147,7 @@ abstract class VWindow(override var dBContext: DBContext = ApplicationContext.ge
     val listeners = modelListener.listenerList
     var i = listeners.size - 2
     while (i >= 0) {
-      if (listeners[i] === VActionListener::class.java) {
+      if (listeners[i] == VActionListener::class.java) {
         (listeners[i + 1] as VActionListener).performAsyncAction(action)
       }
       i -= 2
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:153]: Change while loops to for Loops for a better readability

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 2]
```diff
@@ -214,7 +218,7 @@ abstract class VWindow(override var dBContext: DBContext = ApplicationContext.ge
     val listeners = modelListener.listenerList
     var i = listeners.size - 2
     while (i >= 0) {
-      if (listeners[i] === MessageListener::class.java) {
+      if (listeners[i] == MessageListener::class.java) {
         var value = (listeners[i + 1] as MessageListener).ask(message, yesIsDefault)
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:222]: Change while to for loop

File: galite-core/src/main/kotlin/org/kopi/galite/preview/VPreviewWindow.kt [hunk 3]
```diff
@@ -18,6 +18,378 @@
...
+  override fun getType(): Int = Constants.MDL_PREVIEW
+
+  /**
+   * The user want to show an help
+   */
+  fun preview(printJob: PrintJob, command: String) {
+    val tempFile: File = Utils.getTempFile("PREVIEW", "JPG")
+    this.printJob = printJob
+    numberOfPages = printJob.numberOfPages
+    printFile = printJob.dataFile
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/preview/VPreviewWindow.kt:79]: In printJob it's var datafile , so we need to change this one or change in printJob

File: galite-core/src/main/kotlin/org/kopi/galite/preview/VPreviewWindow.kt [hunk 4]
```diff
@@ -18,6 +18,378 @@
...
+  // ---------------------------------------------------------------------
+  var currentPage = 0
+    private set
+  var numberOfPages = 0
+    private set
+  private var command: String? = null
+  var printJob: PrintJob? = null
+    private set
+  private var printFile: File? = null
+  private var imageFile: String? = null
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/preview/VPreviewWindow.kt:299]: Verify nullability ( there are no != null conditions on them in kopi )

Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/preview/VPreviewWindow.kt:299]: Not nullable

File: galite-core/src/main/kotlin/org/kopi/galite/preview/VPreviewWindow.kt [hunk 5]
```diff
@@ -18,6 +18,378 @@
...
+    setActorEnabled(CMD_LEFT, currentPage > 1)
+    setActorEnabled(CMD_RIGHT, currentPage < numberOfPages)
+    setActorEnabled(CMD_LAST, currentPage < numberOfPages)
+  }
+
+  fun getPreviewFileName(current: Int): String {
+    return "$imageFile$current.JPG"
+  }
+
+  override fun getTitle(): String? {
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/preview/VPreviewWindow.kt:284]: not nullable

**Review by ZMarouani — CHANGES_REQUESTED** [review]

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 6]
```diff
@@ -47,7 +47,7 @@ abstract class VWindow(override var dBContext: DBContext = ApplicationContext.ge
   private var extraTitle: String? = null
   private var display: UWindow? = null
   private var actors: ArrayList<VActor> = arrayListOf()
```
[the remark this answers is not in the document]
Comment (hfazai) [reply galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:49]: actors is gonna be intialized anyway so i think it's better to leave it not nullable

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 7]
```diff
@@ -214,7 +218,7 @@ abstract class VWindow(override var dBContext: DBContext = ApplicationContext.ge
     val listeners = modelListener.listenerList
     var i = listeners.size - 2
     while (i >= 0) {
-      if (listeners[i] === MessageListener::class.java) {
+      if (listeners[i] == MessageListener::class.java) {
         var value = (listeners[i + 1] as MessageListener).ask(message, yesIsDefault)
         when (value) {
```
[the remark this answers is not in the document]
Comment (hfazai) [reply galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:224]: for better visibility I think we better keep it as it is

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 8]
```diff
@@ -109,7 +109,11 @@ abstract class VWindow(override var dBContext: DBContext = ApplicationContext.ge
   /**
    * Resets form to initial state
    */
-  abstract fun reset()
+  fun reset() {
+    // do nothing
+
+    // TODO set it abstract
```
[the remark this answers is not in the document]
Comment (hfazai) [reply galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:116]: I kept it as it is in kopi. I also mark it //TODO
when we finsh model we'll see what we can do for it

**Review by ZMarouani — APPROVED** [review]

**Pull request merged**